### ISSUE TITLE
This module should be documentation instead (and perhaps a BC wrapper in core)

### ISSUE THREAD

**Issue by jenlampton** [backdrop-contrib/entity_metadata_wrapper#6]
An entity metadata wrapper is not necessary in Backdrop. The only reason this module would be needed, is if the module that depends on it was not properly ported.

It's likely that's happened because it's too fricking hard to figure out what to do instead!  We need to do two things before we can remove this contrib module:

* Properly document what to do instead, with real, workable examples.
* See if there's any way we can introduce a backwards-compatible solution directly into core, preferably one that will properly log to watchdog when modules are still using entity metadata wrappers.

**Comment by jenlampton** [comment]
Hm... I'm thinking it might also be a best to delete any "official" releases of this module. Since it only exists to fix problems that need to be addressed in other modules, it's really only a dev tool. What do others think about that?

**Comment by docwilmot** [comment]
I agree this module being a requirement of Ubercart is a bit strange TBH and I think it needs some decision. 
But it started as a viable project. It was really a port of the metadata wrapper part of CTools only. We made the (probably wrong) decision to make Rules dependent on it since many of the functions it contained were required by Rules, but its morphed a bit since. I am seeing it a bit now like CTools was: a collection of code other modules use to get things done.

**Comment by jenlampton** [comment]
> It was really a port of the metadata wrapper part of CTools only.

Yes, and that's the part we don't need, because that is in core already (though different - it just needs to be documented, with working examples, etc...)

> I am seeing it a bit now like CTools was: a collection of code other modules use to get things done.

It sounds like all the things that are not "entity metadata wrapper" should go into a module that doesn't have that name, since we don't want people thinking they need (or should use) an entity metadata wrapper in Backdrop. It would make things very confusing when that part module was removed, if all the other parts that were unrelated, remained. 

Or maybe we rename this whole project to "Entity Plus" or something, so we can safely remove the entity metadata wrapper bit later on?

**Comment by docwilmot** [comment]
> that is in core already

Not exactly. https://github.com/backdrop/backdrop-issues/issues/461#issuecomment-67208290

But I agree better terminolgy might be needed. This is essentially Entity API contrib nearly verbatim. Perhaps a new Entity Extended module, and have Rules, UC, as well as the current EMW be dependednt on that.

**Comment by oadaeh** [comment]
I think it would be very handy to have those things documented, as I am getting ready to port a few things over that are heavily dependent on Drupal's Entity API module, and I was event going to create an Entity API contrib module to provide for all the things that I perceived as not being handled in Backdrop.

**Comment by docwilmot** [comment]
@jenlampton https://github.com/backdrop-contrib/entity_plus.

**Comment by laryn** [comment]
@docwilmot Would it make sense to update README to say "Don't install this, use `entity_plus`, similar to what you put in the E+ README:

> If your module depends on the module `Entity Metadata Wrapper` in Backdrop contrib, read the following and take steps to change the dependency to [this module](https://github.com/backdrop-contrib/entity_plus) instead.

Then perhaps even wrap everything in EMW in a `if (!module_exists('entity_plus') { ... }` if possible so they can be installed side-by-side during transitions from EMW to E+? I am not sure if that will make it smoother, it's just an idea.

I've started branches for both Rules and Ubercart to try to move them to E+ instead of EMW but they both might need some work yet and it's difficult to swap out EMW and E+ (in my test site I end up activating E+, pulling the new branches with E+ dependency, and truncating the `cache_bootstrap` table in the database, which is not great for general upgrade instructions...)

 - https://github.com/laryn/rules/tree/EntityPlus
 - https://github.com/laryn/ubercart/tree/EntityPlus

**Comment by docwilmot** [comment]
Yes thats a great idea. 
This is working out to be so much more trouble than I expected it would be.

**Comment by laryn** [comment]
@docwilmot See what you think:
https://github.com/backdrop-contrib/entity_metadata_wrapper/pull/8

This problem is starting to crop up in the wild: https://github.com/backdrop-contrib/entityreference/issues/17
